### PR TITLE
Save identification and maintainance (I&M) data to nonvolatile memory

### PR DESCRIPTION
**Pull request by pyhys**

### CODE REVIEW

File: include/pnet_api.h [hunk 1]
```diff
@@ -936,7 +937,7 @@ typedef struct pnet_im_2
  * The I&M3 data record is read-write by the controller.
  *
  * This data record is optional. If this data record is supported
- * by the application then bit 2 in the im_supported member of I&M0 shall be set.
+ * by the application then bit 3 in the im_supported member of I&M0 shall be set.
```
Comment (fredrikdanielmoller) [inline include/pnet_api.h:951]: Documentation for pnet_im_4_t also needs an update

File: src/device/pf_fspm.c [hunk 2]
```diff
@@ -376,26 +388,32 @@ int pf_fspm_cm_write_ind(
          pf_get_im_1(&get_info, &pos, &net->fspm_cfg.im_1_data);
          if ((get_info.result == PF_PARSE_OK) && (pos == write_length))
          {
+            LOG_INFO(PNET_LOG, "FSPM(%d): PLC is writing I&M01 data. Function: %s Location: %s\n",
```
Comment (fredrikdanielmoller) [inline src/device/pf_fspm.c:493]:
```suggestion
            LOG_INFO(PNET_LOG, "FSPM(%d): PLC is writing I&M1 data. Function: %s Location: %s\n",
```

File: src/device/pf_fspm.c [hunk 3]
```diff
@@ -376,26 +388,32 @@ int pf_fspm_cm_write_ind(
...
+                __LINE__, net->fspm_cfg.im_1_data.im_tag_function, net->fspm_cfg.im_1_data.im_tag_location);
             ret = 0;
          }
          else
          {
+            LOG_ERROR(PNET_LOG, "FSPM(%d): Wrong length of incoming I&M01 data\n", __LINE__);
```
Comment (fredrikdanielmoller) [inline src/device/pf_fspm.c:499]:
```suggestion
            LOG_ERROR(PNET_LOG, "FSPM(%d): Wrong length of incoming I&M1 data\n", __LINE__);
```

File: src/device/pf_fspm.c [hunk 4]
```diff
@@ -376,26 +388,32 @@ int pf_fspm_cm_write_ind(
...
          }
          break;
       case PF_IDX_SUB_IM_2:
+
          /* Do not count the terminator byte */
          if (write_length == (sizeof(net->fspm_cfg.im_2_data) - 1))
          {
             memcpy(&net->fspm_cfg.im_2_data, p_write_data, sizeof(net->fspm_cfg.im_2_data) - 1);
             net->fspm_cfg.im_2_data.im_date[sizeof(net->fspm_cfg.im_2_data) - 1] = '\0';
+            LOG_INFO(PNET_LOG, "FSPM(%d): PLC is writing I&M02 data. Date: %s\n", __LINE__, net->fspm_cfg.im_2_data.im_date);
```
Comment (fredrikdanielmoller) [inline src/device/pf_fspm.c:513]:
```suggestion
            LOG_INFO(PNET_LOG, "FSPM(%d): PLC is writing I&M2 data. Date: %s\n", __LINE__, net->fspm_cfg.im_2_data.im_date);
```

File: src/device/pf_fspm.c [hunk 5]
```diff
@@ -376,26 +388,32 @@ int pf_fspm_cm_write_ind(
...
             ret = 0;
          }
          else
          {
+            LOG_ERROR(PNET_LOG, "FSPM(%d): Wrong length of incoming I&M02 data\n", __LINE__);
```
Comment (fredrikdanielmoller) [inline src/device/pf_fspm.c:518]:
```suggestion
            LOG_ERROR(PNET_LOG, "FSPM(%d): Wrong length of incoming I&M2 data\n", __LINE__);
```

File: src/device/pf_fspm.c [hunk 6]
```diff
@@ -408,10 +426,12 @@ int pf_fspm_cm_write_ind(
          {
             memcpy(&net->fspm_cfg.im_3_data, p_write_data, sizeof(net->fspm_cfg.im_3_data) - 1);
             net->fspm_cfg.im_3_data.im_descriptor[sizeof(net->fspm_cfg.im_3_data) - 1] = '\0';
+            LOG_INFO(PNET_LOG, "FSPM(%d): PLC is writing I&M3 data. Descriptor: %s\n", __LINE__, net->fspm_cfg.im_3_data.im_descriptor);
             ret = 0;
          }
          else
          {
+            LOG_ERROR(PNET_LOG, "FSPM(%d): Wrong length of incoming I&M03 data\n", __LINE__);
```
Comment (fredrikdanielmoller) [inline src/device/pf_fspm.c:536]:
```suggestion
            LOG_ERROR(PNET_LOG, "FSPM(%d): Wrong length of incoming I&M3 data\n", __LINE__);
```

File: src/device/pf_fspm.c [hunk 7]
```diff
@@ -421,18 +441,21 @@ int pf_fspm_cm_write_ind(
       case PF_IDX_SUB_IM_4:
          if (write_length == sizeof(net->fspm_cfg.im_4_data))
          {
+            LOG_INFO(PNET_LOG, "FSPM(%d): PLC is writing I&M04 data\n", __LINE__);
```
Comment (fredrikdanielmoller) [inline src/device/pf_fspm.c:546]:
```suggestion
            LOG_INFO(PNET_LOG, "FSPM(%d): PLC is writing I&M4 data\n", __LINE__);
```

File: src/device/pf_fspm.c [hunk 8]
```diff
@@ -421,18 +441,21 @@ int pf_fspm_cm_write_ind(
...
             memcpy(&net->fspm_cfg.im_4_data, p_write_data, sizeof(net->fspm_cfg.im_4_data));
             ret = 0;
          }
          else
          {
+            LOG_ERROR(PNET_LOG, "FSPM(%d): Wrong length of incoming I&M04 data\n", __LINE__);
```
Comment (fredrikdanielmoller) [inline src/device/pf_fspm.c:552]:
```suggestion
            LOG_ERROR(PNET_LOG, "FSPM(%d): Wrong length of incoming I&M4 data\n", __LINE__);
```

**Review by fredrikdanielmoller — CHANGES_REQUESTED** [review]

File: src/device/pf_fspm.c [hunk 9]
```diff
@@ -63,6 +63,100 @@ int pf_fspm_validate_configuration(
...
+   {
+      LOG_DEBUG(PNET_LOG,"FSPM(%d): Did read I&M settings from nvm.\n", __LINE__);
+      memcpy(&net->fspm_cfg.im_1_data, &file_im.im1, sizeof(pnet_im_1_t));
+      memcpy(&net->fspm_cfg.im_2_data, &file_im.im2, sizeof(pnet_im_2_t));
+      memcpy(&net->fspm_cfg.im_3_data, &file_im.im3, sizeof(pnet_im_3_t));
+      memcpy(&net->fspm_cfg.im_4_data, &file_im.im4, sizeof(pnet_im_4_t));
+   }
+   else
+   {
+      LOG_DEBUG(PNET_LOG,"FSPM(%d): Could not yet read I&M settings from nvm. Use values from user configuration.\n", __LINE__);
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_fspm.c:90]: This is the normal case when starting up the first time. In order to not scare users, I have included "yet".

File: src/device/pf_fspm.c [hunk 10]
```diff
@@ -63,6 +63,100 @@ int pf_fspm_validate_configuration(
...
+{
+   pnet_im_nvm_t           file_im;
+   bool                    save = false;
+   const char              *p_file_directory = NULL;
+
+   (void)pf_cmina_get_file_directory(net, &p_file_directory);
+
+   if (pf_file_load(p_file_directory, PNET_FILENAME_IM, &file_im, sizeof(pnet_im_nvm_t)) == 0)
+   {
+      if (memcmp(&file_im.im1, &net->fspm_cfg.im_1_data, sizeof(pnet_im_1_t)) != 0)
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_fspm.c:114]: Good idea, I will do that.

File: src/device/pf_cmwrr.c [hunk 11]
```diff
@@ -258,6 +266,7 @@ int pf_cmwrr_rm_write_ind(
    p_write_result->add_data_1 = p_result->add_data_1;
    p_write_result->add_data_2 = p_result->add_data_2;
 
+   /* Restart timer */
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_cmwrr.c:269]: Changed to "Update the CMSM timer, which monitors start-up timeout"

File: include/pnet_api.h (hunk 1 above)
Comment (pyhys) [reply include/pnet_api.h:951]: Done

File: src/device/pf_fspm.c (hunk 5 above)
Comment (pyhys) [reply src/device/pf_fspm.c:518]: Done

File: src/device/pf_fspm.c (hunk 6 above)
Comment (pyhys) [reply src/device/pf_fspm.c:536]: Done

File: src/device/pf_fspm.c (hunk 7 above)
Comment (pyhys) [reply src/device/pf_fspm.c:546]: Done

File: src/device/pf_fspm.c (hunk 8 above)
Comment (pyhys) [reply src/device/pf_fspm.c:552]: Done

File: src/device/pf_fspm.c (hunk 4 above)
Comment (pyhys) [reply src/device/pf_fspm.c:513]: Done

File: src/device/pf_fspm.c (hunk 2 above)
Comment (pyhys) [reply src/device/pf_fspm.c:493]: Done

File: src/device/pf_fspm.c (hunk 3 above)
Comment (pyhys) [reply src/device/pf_fspm.c:499]: Done

File: sample_app/sampleapp_common.c [hunk 12]
```diff
@@ -758,11 +758,14 @@ int app_adjust_stack_configuration(
    stack_config->im_0_data.im_profile_specific_type = 0x5678;
    stack_config->im_0_data.im_version_major = 1;
    stack_config->im_0_data.im_version_minor = 1;
-   stack_config->im_0_data.im_supported = 0x001e;        /* Only I&M0..I&M4 supported */
-   strcpy(stack_config->im_1_data.im_tag_function, "");
-   strcpy(stack_config->im_1_data.im_tag_location, "");
-   strcpy(stack_config->im_2_data.im_date, "");
-   strcpy(stack_config->im_3_data.im_descriptor, "");
+   stack_config->im_0_data.im_supported = PNET_SUPPORTED_IM1 |
```
[the remark this answers is not in the document]
Comment (pyhys) [reply sample_app/sampleapp_common.c:761]: Now verifies `.im_supported` and also adjusts the behavior when the PLC tries to read and write.

File: src/device/pf_fspm.c [hunk 13]
```diff
@@ -63,6 +63,100 @@ int pf_fspm_validate_configuration(
...
+      }
+
+      if (save == true)
+      {
+         LOG_DEBUG(PNET_LOG,"FSPM(%d): Updating nvm stored I&M settings.\n", __LINE__);
+      }
+   }
+   else
+   {
+      LOG_DEBUG(PNET_LOG,"FSPM(%d): First nvm saving of I&M settings.\n", __LINE__);
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_fspm.c]: Done

File: src/device/pf_fspm.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_fspm.c]: Done

File: src/device/pf_fspm.c [hunk 14]
```diff
@@ -63,6 +63,100 @@ int pf_fspm_validate_configuration(
...
+         save = true;
+      }
+      else if (memcmp(&file_im.im4, &net->fspm_cfg.im_4_data, sizeof(pnet_im_4_t)) != 0)
+      {
+         save = true;
+      }
+
+      if (save == true)
+      {
+         LOG_DEBUG(PNET_LOG,"FSPM(%d): Updating nvm stored I&M settings.\n", __LINE__);
```
[the remark this answers is not in the document]
Comment (pyhys) [reply src/device/pf_fspm.c]: Done

**Review by olbjo — APPROVED** [review]

File: src/device/pf_fspm.c [hunk 15]
```diff
@@ -60,9 +62,94 @@ int pf_fspm_validate_configuration(
...
+ *
+ * Compares with the content of already stored settings (in order not to
+ * wear out the flash chip)
+ *
+ * @param net              InOut: The p-net stack instance
+ */
+static void pf_fspm_save_im_if_modified(
+   pnet_t                  *net)
+{
+   pnet_im_nvm_t           output_im;
```
Comment (fredrikdanielmoller) [inline src/device/pf_fspm.c:114]:
```suggestion
   pnet_im_nvm_t           output_im = {0};
```
(ensures that padding bytes are initialised as needed for subsequent memcmp())

**Review by fredrikdanielmoller — CHANGES_REQUESTED** [review]

File: src/device/pf_fspm.c (hunk 15 above)
Comment (pyhys) [reply src/device/pf_fspm.c]: Done (also initialized current_file_im to be consistent)

**Review by fredrikdanielmoller — APPROVED** [review]

**Pull request merged**